File: advent_2015/src/python/advent_2015_day_03.py
with open('day3_input.txt') as f:
    lines = f.readlines()

input_str = lines[0]

# part1
house_set = set()
pos_x = 0
pos_y = 0

for direction in input_str:
    if direction == "^":
        pos_y+=1
    elif direction == "v":
        pos_y-=1
    elif direction == ">":
        pos_x+=1
    elif direction == "<":
        pos_x-=1
        
    house_set.add((pos_x,pos_y))

print(len(house_set))

# part2
house_set_2 = set()
pos_santa_x = 0
pos_santa_y = 0
pos_robo_x = 0
pos_robo_y = 0

for turn, direction in enumerate(input_str):
    increment_x = increment_y = 0
    if direction == "^":
        increment_y += 1
    elif direction == "v":
        increment_y -= 1
    elif direction == ">":
        increment_x += 1
    elif direction == "<":
        increment_x -= 1
    
    if turn % 2 == 0:
        pos_santa_x += increment_x
        pos_santa_y += increment_y
        house_set_2.add((pos_santa_x,pos_santa_y))
    else:
        pos_robo_x += increment_x
        pos_robo_y += increment_y
        house_set_2.add((pos_robo_x, pos_robo_y))

print(len(house_set_2))

File: advent_2015/src/bin/advent_2015_day_03.rs
use advent_2015::file_input;
use std::collections::HashSet;

fn part_1(input_str: &str) -> usize {
    let mut pos_x = 0;
    let mut pos_y = 0;
    let mut visited_houses = HashSet::new();

    for direction in input_str.chars() {
        match direction {
            '^' => pos_y += 1,
            'v' => pos_y -= 1,
            '>' => pos_x += 1,
            '<' => pos_x -= 1,
            _ => (),
        };

        visited_houses.insert((pos_x, pos_y));
    }

    visited_houses.len()
}

fn part_2(input_str: &str) -> usize {
    let mut pos_santa_x = 0;
    let mut pos_santa_y = 0;
    let mut pos_robo_x = 0;
    let mut pos_robo_y = 0;
    let mut visited_houses = HashSet::new();

    for (turn, direction) in input_str.chars().enumerate() {
        let mut increment_y = 0;
        let mut increment_x = 0;
        match direction {
            '^' => increment_y += 1,
            'v' => increment_y -= 1,
            '>' => increment_x += 1,
            '<' => increment_x -= 1,
            _ => (),
        };

        if turn % 2 == 0 {
            pos_santa_x += increment_x;
            pos_santa_y += increment_y;
            visited_houses.insert((pos_santa_x, pos_santa_y));
        } else {
            pos_robo_x += increment_x;
            pos_robo_y += increment_y;
            visited_houses.insert((pos_robo_x, pos_robo_y));
        }
    }

    visited_houses.len()
}

fn main() {
    let lines = file_input::read_input_file();
    let input_str: &str = &lines[0];
    println!("{}", part_1(input_str));
    println!("{}", part_2(input_str));
}
